napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,146 @@
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        if existing_task := self._find_existing_task(layers):
+            logger.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
+
+        # store task for cancellation logic
+        # this is purposefully done before adding the done callback to ensure
+        # that the task is added before the done callback can be executed
+        with self._lock_layers_to_task:
+            self._layers_to_task[tuple(requests.keys())] = task
+
+        task.add_done_callback(self._on_slice_done)
+
+        return task
+
+    def shutdown(self) -> None:
+        """Should be called from the main thread when this is no longer needed."""
+        self._executor.shutdown()
+
+    def _slice_layers(self, requests: Dict) -> Dict:
+        """Iterates through a dictionary of request objects and call the slice
+        on each individual layer. Can be called from the main or slicing thread."""
+        return {
+            layer: layer._get_slice(request)
+            for layer, request in requests.items()
+        }
+
+    def _on_slice_done(self, task: Future[Dict]) -> None:
+        """This is the "done_callback" which is added to each task.
+        Can be called from the main or slicing thread.
+        """
+        if not self._try_to_remove_task(task):
+            logger.debug('Task not found')
+            return
+
+        if task.cancelled():
+            logger.debug('Cancelled task')
+            return
+        result = task.result()
+        self.events.ready(Event('ready', value=result))
+
+    def _try_to_remove_task(self, task) -> bool:
+        """Attempt to remove task, return false if task not found, return true
+        if task removed from layers_to_task dict"""
+        with self._lock_layers_to_task:
+            layers = None
+            for k_layers, v_task in self._layers_to_task.items():
+                if v_task == task:
+                    layers = k_layers
+                    break
+
+            if not layers:
+                return False
+            del self._layers_to_task[layers]
+        return True
+
+    def _find_existing_task(self, layers) -> Optional[Future[Dict]]:
```

Comment:
Type annotation? 

In general, I do not understand what these functions should do. 
After reading the code, it returns a task connected to a subset of provided layers. 
But I'm not sure if this should work way. A docstring with a description will be nice.

Reply:
I've added a docstring.